resolwe/flow/managers/listener.py

```diff
@@ -73,9 +85,103 @@
 logger = logging.getLogger(__name__)
 
 
-class PythonProcessMixin:
+class ListenerPlugin(metaclass=abc.ABCMeta):
+    """Base class for listener plugin classes."""
+
+    @classmethod
+    def __init_subclass__(cls: Type["ListenerPlugin"], **kwargs):
+        """Register class with the registry on initialization."""
+        super().__init_subclass__(**kwargs)
+        ListenerPlugins().add_plugin(cls)
+
+    def _get_handler_name(self, command_name: str) -> str:
+        """Get the handler name."""
+        return f"handle_{command_name}"
+
+    def get_handler(
+        self, command_name: str
+    ) -> Optional[Callable[[Message, "Processor"], Response]]:
+        """Return the plugin handler
+
+        When so such plugin exists return None.
+        """
+        return getattr(self, self._get_handler_name(command_name), None)
+
+
+class ListenerPlugins(MutableMapping[str, object]):
+    """Class for registering listener plugins."""
+
+    __instance = None  #  A single instance of this class
+    _plugin_classes: Dict[str, Type[ListenerPlugin]] = dict()
+    _plugins: Dict[str, ListenerPlugin] = dict()
+
+    @classmethod
+    def __new__(cls, *args, **kwargs):
+        """Create or return singleton."""
+        if ListenerPlugins.__instance is None:
+            ListenerPlugins.__instance = super().__new__(cls)
+
+        return ListenerPlugins.__instance
+
+    def _get_class_name(self, plugin_class: Type[ListenerPlugin]) -> str:
+        """Get the full class name."""
+        return "{}.{}".format(plugin_class.__module__, plugin_class.__name__)
+
+    def add_plugin(self, plugin_class: Type[ListenerPlugin]):
+        """Add new plugin class.
+
+        :raises RuntimeError: when plugin is already registered.
+        """
+        plugin_name = self._get_class_name(plugin_class)
+        assert (
+            plugin_name not in self._plugin_classes
+        ), "Plugin named {plugin_name} already registered."
+        self._plugin_classes[plugin_name] = plugin_class
+
+    def get_handler(self, command_name: str) -> Optional[Callable[[Message], Response]]:
+        """Return the handler for the given command.
+
+        When such handler exists in multiple plugins one is chosen at random.
+        """
+        for plugin_name in self._plugin_classes:
+            if plugin_name not in self._plugins:
+                self._plugins[plugin_name] = self._plugin_classes[plugin_name]()
+            handler = self._plugins[plugin_name].get_handler(command_name)
+            if handler is not None:
+                return handler
+        return None
+
+    def __getitem__(self, item):
```

Comment:
What is the use-case for getter and setter? I see this as an alternative to `get_handler` and `add_plugin` methods. Or does it make sense to use both of them?

Reply:
I am actually deciding on this. Which approach do you prefer?

Reply:
Both are ok for me, but if I would have to choose one, I would go with "Explicit is better than implicit."

Reply:
OK. I have removed the mapping methods and kept explicit methods.